src/sentry/search/events/builder/metrics.py

```diff
@@ -772,6 +774,53 @@ def resolve_limit(self, limit: Optional[int]) -> Optional[Limit]:
     def resolve_granularity(self) -> Granularity:
         return Granularity(self._granularity)
 
+    def _get_snql_using_metrics_layer(self, snuba_request: Request) -> Request:
+        # We use directly the query builder which is not a good practice, considering that the metrics layer
+        # should hide snql generation.
+        from sentry.snuba.metrics import SnubaQueryBuilder
+        from sentry.snuba.metrics.mqb_query_transformer import transform_mqb_query_to_metrics_query
+
+        snuba_queries, _ = SnubaQueryBuilder(
+            projects=self.params.projects,
+            metrics_query=transform_mqb_query_to_metrics_query(
+                snuba_request.query, is_alerts_query=True
+            ),
+            use_case_id=UseCaseKey.PERFORMANCE
+            if self.is_performance
+            else UseCaseKey.RELEASE_HEALTH,
+        ).get_snuba_queries()
+
+        if len(snuba_queries) == 0 or len(snuba_queries) > 1:
+            raise NotImplementedError(
+                "get_snql_query cannot be implemented for MetricsQueryBuilder"
+            )
+
+        # We take only the first query, supposing a single query is generated.
+        entity = list(snuba_queries.keys())[0]
+        snuba_request.query = snuba_queries[entity]["totals"]
+
+        return snuba_request
+
+    def run_query(self, referrer: str, use_cache: bool = False) -> Any:
+        if self.use_metrics_layer:
+            # We have to copy the entire code of the run query here, because of how the get_snql_query() method
+            # behaves. The get_snql_query() of MetricsQueryBuilder returns a snql query that is understood by the mqb
+            # transformer. If we were to call get_snql_query() of this class, we would already get snql generated
+            # by the metrics layer, which is not understood by the transformer.
```

Comment:
Isn't this only true because you're calling `super().get_snql_query()`? Wouldn't you not have to do this at all if you didn't overwrite `run_query` since you're overwriting `get_snql_query` already?

To add to this, do we need the `_get_snql_using_metrics_layer` helper function? Can't we just overwrite get_snql_query directly instead?

Reply:
We could do that sure, I just wanted to reduce code duplication.

Reply:
In your diff, you remove the `run_query` function which will mean we are going to use the `run_query` of the parent class that will use the `get_snql_query` of the child class during execution. All of my code was to circumvent this issue.

Reply:
If i understand what you're trying to do better this may be a more accurate diff for what you're doing: https://gist.github.com/wmak/d654c19be1fb1304a340f9310b0ca115

The idea is to introduce a new `get_metrics_layer_query` that returns a Request compatible with the metric layer, return `NotImplemented` on the base `MetricsQueryBuilder.get_snql_query`. And only implement `get_snql_query` on the `AlertQueryBuilder`

Reply:
This approach could work but we then would simply need to implement`get_metrics_layer_query` in the `MetricsQueryBuilder` as that building process is not different from the alerts one and this would converge to an implementation similar to mine.